facebook/react-native
# Add ability to control scroll animation duration for Android

## Motivation

This is one of the more sought after feature requests for RN:
https://react-native.canny.io/feature-requests/p/add-speed-attribute-to-scrollto

This PR adds the support to add a "duration" whenever using "scrollTo" or "scrollToEnd" with
a scrollView. Currently this only exists for Android as the iOS implementation will be somewhat more involved.

This PR is also backwards compatible and does not yet deprecate the "animated" boolean. It may not make sense to ever deprecate "animated", as it could be the flag that is used when devs want the system default duration (which is 250ms for Android). I'm not sure what it is for iOS. It would simplify things to remove "animated", though.

## Test Plan

1. Create a ScollView component that overflows the bounds of the container, e.g.:

```javascript
  render() {
    return (
      <View style={styles.container}>
        <Button onPress={this.scrollToPosition} title="Scroll to Position" color="#841584" />
        <ScrollView style={styles.scrollView} ref={scrollView => this.scrollView = scrollView}>
          {items.map(item => {
            return <View key={item.key} style={styles.scrollViewItem}><Text>{item.text}</Text></View>
          })}
        </ScrollView>
      </View>
    );
  }
```

2. Using the "scrollView" ref and the Button callback, scroll to a position with a desired duration speed, e.g:

```javascript
scrollToPosition = () => {
  this.scrollView && this.scrollView.scrollTo({
    y: 500
    duration: 1000
  });
};
```

3. Ensure the legacy "animated" boolean works as expected:

```javascript
scrollToPosition = () => {
  this.scrollView && this.scrollView.scrollTo({
    y: 500
    animated: true
  });
};
```

4. Test that you can animate the x position as well (change the ScrollView to be horizontal).
5. Test that the scroll animation is canceled when touching on the scroll view in the middle of an animation.
6. Test that repeated calls to animate the scrollView don't queue up (e.g. smashing the <Button> in the example above, and then tapping into the scrollView should cancel any animation movement).


Vertical scroll:
https://drive.google.com/file/d/1VhaolHx63GOv0YWTg7ay0GtzW1nIk635/view?usp=sharing

Horizontal scroll:
https://drive.google.com/file/d/1yq5P1k4-KPPu9BhS6B9liL-NRGX1Z4gA/view?usp=sharing


## Related PRs

Abandoned PR to bring "duration" to iOS:
https://github.com/facebook/react-native/pull/1334

Updated docs on react-native-website:
https://github.com/facebook/react-native-website/pull/115

## Release Notes

 [ANDROID] [ENHANCEMENT] [ScrollView] - Added support for controlling animation duration speed for methods "scrollTo" and "scrollToEnd"
